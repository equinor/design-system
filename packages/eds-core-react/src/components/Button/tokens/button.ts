import { tokens } from '@equinor/eds-tokens'
import { ButtonToken } from '../Button.types'

const {
  typography: {
    navigation: { button: buttonTypography },
  },
  colors: {
    interactive: {
      disabled__text: { rgba: disabledTextColor },
      disabled__border: { rgba: disabledBorderColor },
      disabled__fill: { rgba: disabledColor },
    },
  },
  spacings: {
    comfortable: { medium },
  },
  shape: {
    button: { minHeight: buttonHeight, borderRadius: buttonBorderRadius },
    _modes: {
      compact: {
        button: { minHeight: compactButtonHeight },
      },
    },
  },
  interactions: { focused: outline },
  clickbounds: {
    default__base: clicboundHeight,
    compact__standard: compactClickboundHeight,
  },
} = tokens

export const button: ButtonToken = {
  background: 'transparent',
  height: `var(--eds_button__height, ${buttonHeight})`,
  typography: {
    ...buttonTypography,
    textAlign: 'center',
    fontSize: `var(--eds_button__font_size, ${buttonTypography.fontSize})`,
  },
  border: {
    type: 'border',
    width: `var(--eds_button__border_width, 1px)`,
    color: 'transparent',
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
    style: 'solid',
  },
  spacings: {
    top: 'var(--eds_button__padding_y, 0)',
    bottom: 'var(--eds_button__padding_y, 0)',
    left: `var(--eds_button__padding_x, ${medium})`,
    right: `var(--eds_button__padding_x, ${medium})`,
  },
  clickbound: {
    height: clicboundHeight,
    width: '100%',
    offset: {
      top: `${(parseInt(clicboundHeight) - parseInt(buttonHeight)) / 2 + 1}px`,
      left: '0',
    },
  },
  entities: {
    icon: {
      height: 'var(--eds_button__icon__size, 24px)',
      width: 'var(--eds_button__icon__size, 24px)',
    },
  },
  states: {
    hover: {
      border: {
        type: 'border',
        width: '1px',
        color: 'transparent',
        radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
        style: 'solid',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        offset: '2px',
        style: 'dashed',
        color: outline.color,
        width: outline.width,
      },
    },
    disabled: {
      background: disabledColor,
      border: {
        type: 'border',
        width: '1px',
        color: disabledBorderColor,
        style: 'solid',
      },
      typography: {
        ...buttonTypography,
        color: disabledTextColor,
        textAlign: 'center',
      },
    },
  },
  modes: {
    compact: {
      height: `var(--eds_button__height_compact, ${compactButtonHeight})`,
      spacings: {
        top: 'var(--eds_button__padding_y_compact, 0)',
        bottom: 'var(--eds_button__padding_y_compact, 0)',
      },
      clickbound: {
        height: compactClickboundHeight,
        width: '100%',
        offset: {
          top: `${
            (parseInt(compactClickboundHeight) -
              parseInt(compactButtonHeight)) /
              2 +
            1
          }px`,
          left: '0',
        },
      },
    },
  },
}
