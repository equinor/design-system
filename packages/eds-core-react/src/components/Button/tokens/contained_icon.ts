import { tokens } from '@equinor/eds-tokens'
import { mergeDeepRight } from 'ramda'

import {
  primary as primaryContained,
  secondary as secondaryContained,
  danger as dangerContained,
} from './contained'
import { ButtonToken } from '../Button.types'

const {
  clickbounds: {
    default__base: clicboundHeight,
    compact__standard: compactClickboundHeight,
  },
  shape,
} = tokens

const contained_icon = {
  height: shape.icon_button.minHeight,
  width: shape.icon_button.minWidth,
  border: {
    width: '0px',
    radius: '50%',
  },
  spacings: { left: '0', right: '0' },
  clickbound: {
    width: clicboundHeight,
    offset: {
      top: `${
        (parseInt(clicboundHeight) - parseInt(shape.icon_button.minWidth)) / 2
      }px`,
      left: `${
        (parseInt(clicboundHeight) - parseInt(shape.icon_button.minWidth)) / 2
      }px`,
    },
  },
  states: {
    hover: {
      border: {
        width: '0px',
        radius: '50%',
      },
    },
  },
  modes: {
    compact: {
      height: shape._modes.compact.icon_button.minHeight,
      width: shape._modes.compact.icon_button.minWidth,
      clickbound: {
        width: compactClickboundHeight,
        offset: {
          top: '0',
          left: `${
            (parseInt(compactClickboundHeight) -
              parseInt(shape._modes.compact.icon_button.minWidth)) /
            2
          }px`,
        },
      },
    },
  },
}

export const primary = mergeDeepRight(
  primaryContained,
  contained_icon,
) as ButtonToken

export const secondary = mergeDeepRight(
  secondaryContained,
  contained_icon,
) as Partial<ButtonToken>
export const danger = mergeDeepRight(
  dangerContained,
  contained_icon,
) as Partial<ButtonToken>
