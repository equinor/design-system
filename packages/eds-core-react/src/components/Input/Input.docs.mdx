import { Links, PropsTable, Story } from './../../../.storybook/components'
import { Input } from '.'

# Input

Is intended to use if you need to more flexibility than the wrapped TextField or select components give.

<Links 
    ariaUrl="https://www.w3.org/TR/wai-aria-1.1/#input"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Input/Input.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Story id="inputs-input--introduction" />
<PropsTable />

## Usage

```tsx
import { Input } from '@equinor/eds-core-react'

<Input />
```

## Examples

### Types

This is the native type property and here is a small selection of examples of all the possible types.

<Story id="inputs-input--types" />

### Variants

<Story id="inputs-input--variants" />

### Disabled

<Story id="inputs-input--disabled" />

### Read only

<Story id="inputs-input--read-only" />

### Accessiblity

It's important to link the `Input` to the corresponding label.

<Story id="inputs-input--accessiblity" />

### Compact

Compact `Input` using `EdsProvider`.

<Story id="inputs-input--compact" />

### With Adornments


<Story id="inputs-input--with-adornments" />

### Casted as textarea


<Story id="inputs-input--casted" />
