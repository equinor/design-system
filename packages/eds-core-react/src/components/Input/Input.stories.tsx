import { useState, useEffect } from 'react'
import { Input, InputProps, Label, EdsProvider, Density } from '../..'
import { Story } from '@storybook/react/types-6-0'
import { ComponentMeta } from '@storybook/react'
import { Stack } from './../../../.storybook/components'
import page from './Input.docs.mdx'

export default {
  title: 'Inputs/Input',
  component: Input,
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof Input>

export const Introduction: Story<InputProps> = (args) => <Input {...args} />

export const Types: Story<InputProps> = () => (
  <Stack>
    <div>
      <Label htmlFor="textfield-normal" label="Normal" />
      <Input
        id="textfield-normal"
        placeholder="Placeholder text"
        autoComplete="off"
      />
    </div>
    <div>
      <Label htmlFor="textfield-number" label="Number" />
      <Input
        type="number"
        id="textfield-number"
        placeholder="Placeholder text"
      />
    </div>
    <div>
      <Label htmlFor="textfield-search" label="Search" />
      <Input
        type="search"
        id="textfield-search"
        placeholder="Placeholder text"
      />
    </div>
    <div>
      <Label htmlFor="textfield-password" label="Password" />
      <Input
        type="password"
        id="textfield-password"
        placeholder="Placeholder text"
      />
    </div>
    <div>
      <Label htmlFor="textfield-email" label="Email" />
      <Input type="email" id="textfield-email" placeholder="Placeholder text" />
    </div>
  </Stack>
)

export const Variants: Story<InputProps> = () => (
  <Stack>
    <div>
      <Label htmlFor="textfield-default" label="Default" />
      <Input
        id="textfield-default"
        placeholder="Placeholder text"
        autoComplete="off"
      />
    </div>
    <div>
      <Label htmlFor="textfield-success" label="Success" />
      <Input
        id="textfield-success"
        placeholder="Placeholder text"
        autoComplete="off"
        variant="success"
      />
    </div>
    <div>
      <Label htmlFor="textfield-warning" label="Warning" />
      <Input
        id="textfield-warning"
        placeholder="Placeholder text"
        autoComplete="off"
        variant="warning"
      />
    </div>
    <div>
      <Label htmlFor="textfield-error" label="Error" />
      <Input
        id="textfield-error"
        placeholder="Placeholder text"
        autoComplete="off"
        variant="error"
      />
    </div>
  </Stack>
)

export const Disabled: Story<InputProps> = () => (
  <>
    <Label htmlFor="textfield-disabled" label="Disabled" />
    <Input id="textfield-disabled" placeholder="Placeholder text" disabled />
  </>
)

export const ReadOnly: Story<InputProps> = () => (
  <>
    <Label htmlFor="textfield-readOnly" label="Read only" />
    <Input id="textfield-readOnly" placeholder="Placeholder text" readOnly />
  </>
)
ReadOnly.storyName = 'Read only'

export const Accessiblity: Story<InputProps> = () => {
  // To wrap the input component inside the label element is not yet supported
  return (
    <>
      <Label htmlFor="a11yExample" label="I use the htmlFor prop" />
      <Input type="text" id="a11yExample" />
    </>
  )
}

export const Compact: Story<InputProps> = () => {
  // To wrap the input component inside the label element is not yet supported
  const [density, setDensity] = useState<Density>('comfortable')

  useEffect(() => {
    // Simulate user change
    setDensity('compact')
  }, [density])

  return (
    <EdsProvider density={density}>
      <Label htmlFor="compact" label="Compact" />
      <Input type="text" id="compact" />
    </EdsProvider>
  )
}
