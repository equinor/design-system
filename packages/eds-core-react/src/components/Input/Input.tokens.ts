import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'
import { mergeDeepRight } from 'ramda'

const {
  colors: {
    ui: {
      background__light: { rgba: background },
    },
    text: { static_icons__default, static_icons__tertiary },
    interactive: {
      disabled__text,
      primary__resting,
      danger__resting,
      danger__hover,
      warning__resting,
      warning__hover,
      success__resting,
      success__hover,
    },
  },
  spacings: {
    comfortable: { small, x_small },
  },
  typography,
  shape,
} = tokens

export type InputToken = ComponentToken

export const input: InputToken = {
  minHeight: shape.straight.minHeight,
  background,
  spacings: {
    left: small,
    right: small,
    top: '6px',
    bottom: '6px',
  },
  typography: {
    ...typography.input.text,
    color: static_icons__default.rgba,
  },
  entities: {
    placeholder: {
      typography: {
        color: static_icons__tertiary.rgba,
      },
    },
  },
  states: {
    disabled: {
      typography: {
        color: disabled__text.rgba,
      },
    },
    readOnly: {
      background: 'transparent',
      boxShadow: 'none',
    },
    active: {
      outline: {
        type: 'outline',
        color: 'transparent',
        width: '1px',
        style: 'solid',
        offset: '0px',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        width: '2px',
        color: primary__resting.rgba,
        style: 'solid',
        offset: '0px',
      },
    },
  },
  boxShadow: 'inset 0px -1px 0px 0px ' + static_icons__tertiary.rgba,
  modes: {
    compact: {
      minHeight: shape._modes.compact.straight.minHeight,
      spacings: {
        left: x_small,
        right: x_small,
        top: x_small,
        bottom: x_small,
      },
    },
  },
}

export const error: InputToken = mergeDeepRight(input, {
  boxShadow: 'inset 0px -1px 0px 0px transparent',
  states: {
    active: {
      outline: {
        type: 'outline',
        color: danger__resting.rgba,
        width: '1px',
        style: 'solid',
        offset: '0px',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        width: '2px',
        color: danger__hover.rgba,
        style: 'solid',
        offset: '0px',
      },
    },
  },
})

export const warning: InputToken = mergeDeepRight(input, {
  boxShadow: 'inset 0px -1px 0px 0px transparent',
  states: {
    active: {
      outline: {
        type: 'outline',
        color: warning__resting.rgba,
        width: '1px',
        style: 'solid',
        offset: '0px',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        width: '2px',
        color: warning__hover.rgba,
        style: 'solid',
        offset: '0px',
      },
    },
  },
})

export const success: InputToken = mergeDeepRight(input, {
  boxShadow: 'inset 0px -1px 0px 0px transparent',
  states: {
    active: {
      outline: {
        type: 'outline',
        color: success__resting.rgba,
        width: '1px',
        style: 'solid',
        offset: '0px',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        width: '2px',
        color: success__hover.rgba,
        style: 'solid',
        offset: '0px',
      },
    },
  },
})

export const inputToken = { input, error, warning, success }
