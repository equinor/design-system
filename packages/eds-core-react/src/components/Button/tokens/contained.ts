import { tokens } from '@equinor/eds-tokens'
import { ButtonToken } from '../Button.types'
import { button } from './button'

import { mergeDeepRight } from 'ramda'

const {
  colors: {
    text: {
      static_icons__primary_white: { rgba: primaryWhite },
    },
    interactive: {
      primary__resting: { rgba: primaryColor },
      primary__hover: { rgba: primaryHoverColor },
      secondary__resting: { rgba: secondaryColor },
      secondary__link_hover: { rgba: secondaryHoverColor },
      danger__resting: { rgba: dangerColor },
      danger__hover: { rgba: dangerHoverColor },
    },
  },

  shape: {
    button: { borderRadius: buttonBorderRadius },
  },
} = tokens

export const primary = mergeDeepRight(button, {
  background: primaryColor,
  typography: {
    color: primaryWhite,
  },
  border: {
    type: 'border',
    style: 'solid',
    width: '1px',
    color: primaryColor,
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
  },
  states: {
    hover: {
      background: primaryHoverColor,
    },
  },
}) as ButtonToken
export const secondary = mergeDeepRight(primary, {
  background: secondaryColor,
  border: {
    color: secondaryColor,
  },
  states: {
    hover: {
      background: secondaryHoverColor,
      border: {
        color: secondaryHoverColor,
      },
    },
  },
}) as ButtonToken

export const danger = mergeDeepRight(primary, {
  background: dangerColor,
  border: {
    color: dangerColor,
  },
  states: {
    hover: {
      background: dangerHoverColor,
      border: {
        color: dangerHoverColor,
      },
    },
  },
}) as ButtonToken
