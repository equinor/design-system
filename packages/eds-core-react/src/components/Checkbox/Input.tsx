/* eslint camelcase: "off" */
import { forwardRef, InputHTMLAttributes } from 'react'
import styled, { ThemeProvider } from 'styled-components'
import {
  checkbox,
  checkbox_outline, // eslint-disable-line camelcase
  checkbox_indeterminate, // eslint-disable-line camelcase
} from '@equinor/eds-icons'
import type { IconData } from '@equinor/eds-icons'
import { checkbox as tokens } from './Checkbox.tokens'
import { spacingsTemplate, outlineTemplate, useToken } from '@equinor/eds-utils'
import { useEds } from '../EdsProvider'

type StyledIconPathProps = {
  $icon: IconData
  name: string
}

const StyledPath = styled.path.attrs<StyledIconPathProps>(({ $icon }) => ({
  fillRule: 'evenodd',
  clipRule: 'evenodd',
  d: $icon.svgPathData as string,
}))<StyledIconPathProps>``

type StyledInputProps = {
  $iconSize: number
}

const Input = styled.input.attrs<StyledInputProps>(({ type = 'checkbox' }) => ({
  type,
}))<StyledInputProps>`
  --scale: ${({ theme, $iconSize }) =>
    parseFloat(theme.clickbound.height) / $iconSize};
  appearance: none;
  width: 100%;
  height: 100%;
  margin: 0;
  grid-area: input;
  transform: scale(var(--scale));
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  &:focus {
    outline: none;
  }
  &[data-focus-visible-added]:focus + svg {
    ${({ theme }) => outlineTemplate(theme.states.focus.outline)}
  }
  &:focus-visible + svg {
    ${({ theme }) => outlineTemplate(theme.states.focus.outline)}
  }
  &:not(:checked) ~ svg path[name='checked'] {
    display: none;
  }
  &:not(:checked) ~ svg path[name='not-checked'] {
    display: inline;
  }
  &:checked ~ svg path[name='not-checked'] {
    display: none;
  }
  &:checked ~ svg path[name='checked'] {
    display: inline;
  }
`

const Svg = styled.svg.attrs(({ height, width, fill }) => ({
  name: null,
  xmlns: 'http://www.w3.org/2000/svg',
  height,
  width,
  fill,
}))`
  grid-area: input;
  pointer-events: none;
`

type StyledInputWrapperProps = { disabled: boolean }

const InputWrapper = styled.span<StyledInputWrapperProps>`
  display: inline-grid;
  grid: [input] 1fr / [input] 1fr;
  position: relative;
  isolation: isolate;
  ${({ theme }) => spacingsTemplate(theme.spacings)}
  &::before {
    content: '';
    position: absolute;
    width: ${({ theme }) => theme.width};
    aspect-ratio: 1/1;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    border-radius: 100%;
  }
  @media (hover: hover) and (pointer: fine) {
    > svg {
      z-index: 1;
    }
    &:hover {
      &::before {
        background-color: ${({ disabled }) =>
          disabled ? 'transparent' : tokens.states.hover.background};
      }
    }
  }
`
export type InputProps = {
  /** If true, the checkbox will be disabled */
  disabled?: boolean
  /** If true, the checkbox appears indeterminate. Important! You'll have to
   * set the native element to indeterminate yourself.
   */
  indeterminate?: boolean
} & InputHTMLAttributes<HTMLInputElement>

export const CheckboxInput = forwardRef<HTMLInputElement, InputProps>(
  function CheckboxInput(
    { disabled = false, indeterminate, style, className, ...rest },
    ref,
  ) {
    const { density } = useEds()
    const token = useToken({ density }, tokens)

    const iconSize = 24
    const fill = disabled
      ? tokens.states.disabled.background
      : tokens.background

    const inputWrapperProps = {
      disabled,
      style,
      className,
    }

    const inputProps = {
      ref,
      disabled,
      ['data-indeterminate']: indeterminate,
      ...rest,
    }

    return (
      <ThemeProvider theme={token}>
        <InputWrapper {...inputWrapperProps}>
          <Input $iconSize={iconSize} {...inputProps} />
          {indeterminate ? (
            <Svg
              width={iconSize}
              height={iconSize}
              viewBox={`0 0 ${iconSize} ${iconSize}`}
              fill={fill}
              aria-hidden
            >
              <StyledPath $icon={checkbox_indeterminate} name="indeterminate" />
            </Svg>
          ) : (
            <Svg
              width={iconSize}
              height={iconSize}
              viewBox={`0 0 ${iconSize} ${iconSize}`}
              fill={fill}
              aria-hidden
            >
              <StyledPath $icon={checkbox} name="checked" />
              <StyledPath $icon={checkbox_outline} name="not-checked" />
            </Svg>
          )}
        </InputWrapper>
      </ThemeProvider>
    )
  },
)
