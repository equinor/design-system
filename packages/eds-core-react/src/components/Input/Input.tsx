import {
  forwardRef,
  ReactNode,
  useState,
  useCallback,
  CSSProperties,
  ElementType,
  ComponentPropsWithoutRef,
} from 'react'
import styled, { css } from 'styled-components'
import {
  typographyMixin,
  spacingsTemplate,
  outlineTemplate,
  useToken,
  OverridableComponent,
} from '@equinor/eds-utils'
import { inputToken as tokens } from './Input.tokens'
import type { InputToken } from './Input.tokens'
import type { Variants } from '../types'
import { useEds } from '../EdsProvider'

const Container = styled.div<StyledProps>(({ $token, disabled, readOnly }) => {
  const { states, entities } = $token

  return css`
    --eds-input-adornment-color: ${entities.adornment.typography.color};
    --eds-input-color: ${$token.typography.color};

    position: relative;
    height: ${$token.height};
    width: ${$token.width};
    display: flex;
    flex-direction: row;
    border: none;
    box-sizing: border-box;
    box-shadow: ${$token.boxShadow};
    background: var(--eds-input-background, ${$token.background});
    ${outlineTemplate($token.outline)}

    &:focus-within {
      --eds-input-adornment-color: ${entities.adornment?.states.focus?.outline
        .color};

      box-shadow: none;
      ${outlineTemplate(states.focus.outline)}
    }

    ${disabled &&
    css`
      --eds-input-adornment-color: ${states.disabled.typography.color};
      --eds-input-color: ${states.disabled.typography.color};
      cursor: not-allowed;
      box-shadow: none;
    `}
    ${readOnly &&
    css({
      background: states.readOnly.background,
      boxShadow: states.readOnly.boxShadow,
    })}
    & > input {
      overflow: hidden;
      white-space: nowrap;
      text-overflow: ellipsis;
    }
  `
})

const StyledInput = styled.input<StyledProps>(
  ({ $token, $paddingLeft, $paddingRight }) => {
    return css`
      width: 100%;
      border: none;
      background: transparent;
      ${spacingsTemplate($token.spacings)}
      ${typographyMixin($token.typography)}
      outline: none;

      padding-left: ${$paddingLeft};
      padding-right: ${$paddingRight};

      &::placeholder {
        color: ${$token.entities.placeholder.typography.color};
      }

      &:disabled {
        color: var(--eds-input-color);
        cursor: not-allowed;
      }
    `
  },
)

type AdornmentProps = {
  $token: InputToken
}

const Adornments = styled.div<AdornmentProps>(({ $token }) => {
  return css`
    position: absolute;
    top: ${$token.spacings.top};
    bottom: ${$token.spacings.bottom};
    display: flex;
    align-items: center;
    ${typographyMixin($token.entities.adornment.typography)}
    color: var(--eds-input-adornment-color);
  `
})

const LeftAdornments = styled(Adornments)(
  ({ $token }) => css`
    left: 0;
    padding-left: ${$token.entities.adornment.spacings.left};
  `,
)

const RightAdornments = styled(Adornments)(
  ({ $token }) => css`
    right: 0;
    padding-right: ${$token.entities.adornment.spacings.right};
  `,
)
type StyledProps = {
  $token: InputToken
  $paddingLeft?: string
  $paddingRight?: string
} & Required<Pick<InputProps, 'readOnly' | 'disabled'>>

export type InputProps = {
  /** Placeholder */
  placeholder?: string
  /** Variant */
  variant?: Variants
  /** Disabled state */
  disabled?: boolean
  /** Type */
  type?: string
  /** Read Only */
  readOnly?: boolean
  /** Left adornments */
  leftAdornments?: ReactNode
  /** Right adornments */
  rightAdornments?: ReactNode
  /** Left adornments props */
  leftAdornmentsProps?: ComponentPropsWithoutRef<'div'>
  /** Right adornments props */
  rightAdornmentsProps?: ComponentPropsWithoutRef<'div'>
  /** Manually specify left adornments width. The width will be the dom element width if not defined */
  leftAdornmentsWidth?: number
  /**  Manually specify right adornments width. The width will be the dom element width if not defined */
  rightAdornmentsWidth?: number
  /** Cast the input to another element */
  as?: ElementType
  /**  */
  className?: string
  style?: CSSProperties
}

export const Input: OverridableComponent<InputProps, HTMLInputElement> =
  forwardRef<HTMLInputElement, InputProps>(function Input(
    {
      variant,
      disabled = false,
      type = 'text',
      leftAdornments,
      rightAdornments,
      readOnly,
      className,
      style,
      leftAdornmentsProps,
      rightAdornmentsProps,
      leftAdornmentsWidth,
      rightAdornmentsWidth,
      ...other
    },
    ref,
  ) {
    const inputVariant = tokens[variant] ? tokens[variant] : tokens.input
    const { density } = useEds()
    const _token = useToken({ density }, inputVariant)() as InputToken

    const [rightAdornmentsRef, setRightAdornmentsRef] =
      useState<HTMLDivElement>()
    const [leftAdornmentsRef, setLeftAdornmentsRef] = useState<HTMLDivElement>()

    const token = useCallback((): InputToken => {
      const _leftAdornmentsWidth =
        leftAdornmentsWidth ||
        (leftAdornmentsRef ? leftAdornmentsRef.clientWidth : 0)
      const _rightAdornmentsWidth =
        rightAdornmentsWidth ||
        (rightAdornmentsRef ? rightAdornmentsRef.clientWidth : 0)
      return {
        ..._token,
        spacings: {
          ..._token.spacings,
          left: `${_leftAdornmentsWidth + parseInt(_token.spacings.left)}px`,
          right: `${_rightAdornmentsWidth + parseInt(_token.spacings.right)}px`,
        },
      }
    }, [
      leftAdornmentsWidth,
      leftAdornmentsRef,
      rightAdornmentsWidth,
      rightAdornmentsRef,
      _token,
    ])()

    const inputProps = {
      ref,
      type,
      disabled,
      readOnly,
      $token: token,
      style: {
        resize: 'none',
      },
      ...other,
    }

    const containerProps = {
      disabled,
      readOnly,
      className,
      style,
      $token: token,
    }

    const _leftAdornmentProps = {
      ...leftAdornmentsProps,
      ref: setLeftAdornmentsRef,
      $token: token,
    }
    const _rightAdornmentProps = {
      ...rightAdornmentsProps,
      ref: setRightAdornmentsRef,
      $token: token,
    }

    return (
      // Not using <ThemeProvider> because of cascading styling messing with adornments
      <Container {...containerProps}>
        {leftAdornments ? (
          <LeftAdornments {..._leftAdornmentProps}>
            {leftAdornments}
          </LeftAdornments>
        ) : null}
        <StyledInput
          $paddingLeft={token.spacings.left}
          $paddingRight={token.spacings.right}
          {...inputProps}
        />
        {rightAdornments ? (
          <RightAdornments {..._rightAdornmentProps}>
            {rightAdornments}
          </RightAdornments>
        ) : null}
      </Container>
    )
  })
