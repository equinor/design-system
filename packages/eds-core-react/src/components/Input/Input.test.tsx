/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react'
import 'jest-styled-components'
import '@testing-library/jest-dom'
import { axe } from 'jest-axe'
import styled from 'styled-components'
import { Input } from './Input'
import * as tokens from './Input.tokens'
import { trimSpaces } from '@equinor/eds-utils'

describe('Input', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(<Input value="input value" readOnly />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test when using label', async () => {
    const { container } = render(
      // eslint-disable-next-line jsx-a11y/label-has-associated-control
      <label>
        Label text
        <Input />
      </label>,
    )
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Should pass a11y test when using aria-label', async () => {
    const { container } = render(<Input aria-label="description" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Has correct default value', () => {
    const value = 'Some value'
    render(<Input id="test-value" value={value} readOnly />)
    const inputElement: HTMLInputElement = screen.queryByDisplayValue(value)

    expect(inputElement.value).toBe(value)
  })

  it('Has correct outline on input when variant is success', () => {
    const label = 'success outline on input'
    render(
      <label>
        {label}
        <Input id="test-success" variant="success" />
      </label>,
    )
    // eslint-disable-next-line testing-library/no-node-access
    const inputWrapper = screen.getByLabelText(label).parentElement

    expect(inputWrapper).toHaveStyleRule(
      'outline',
      `${tokens.input.outline.width} solid ${trimSpaces(
        tokens.success.outline.color,
      )}`,
    )
  })

  it('Has correct outline on input when variant is warning', () => {
    const label = 'warning outline on input'
    render(
      <label>
        {label}
        <Input id="test-warning" variant="warning" />
      </label>,
    )
    // eslint-disable-next-line testing-library/no-node-access
    const inputWrapper = screen.getByLabelText(label).parentElement

    expect(inputWrapper).toHaveStyleRule(
      'outline',
      `${tokens.input.outline.width} solid ${trimSpaces(
        tokens.warning.outline.color,
      )}`,
    )
  })

  it('Has correct outline on input when variant is error', () => {
    const label = 'error outline on input'
    render(
      <label>
        {label}
        <Input id="test-error" variant="error" />
      </label>,
    )
    // eslint-disable-next-line testing-library/no-node-access
    const inputWrapper = screen.getByLabelText(label).parentElement

    expect(inputWrapper).toHaveStyleRule(
      'outline',
      `${tokens.input.outline.width} solid ${trimSpaces(
        tokens.error.outline.color,
      )}`,
    )
  })

  const StyledInput = styled(Input)`
    margin-top: 48px;
  `
  it('Can extend the css of the component', () => {
    render(
      <StyledInput
        id="test-css-extend"
        variant="error"
        value="textfield"
        readOnly
      />,
    )

    // eslint-disable-next-line testing-library/no-node-access
    const inputWrapper = screen.getByDisplayValue('textfield').parentElement

    expect(inputWrapper).toHaveStyleRule('margin-top', '48px')
  })
})
