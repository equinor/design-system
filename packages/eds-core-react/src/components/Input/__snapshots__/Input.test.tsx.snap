// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  --eds-input-adornment-color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
  --eds-input-color: var(--eds_text__static_icons__default,rgba(61,61,61,1));
  position: relative;
  height: 36px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border: none;
  box-sizing: border-box;
  box-shadow: inset 0px -1px 0px 0px var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
  background: var(--eds-input-background,var(--eds_ui_background__light,rgba(247,247,247,1)));
  outline: 1px solid transparent;
  outline-offset: 0px;
  background: transparent;
  box-shadow: none;
}

.c0:focus-within {
  box-shadow: none;
  outline: 2px solid var(--eds_interactive_primary__resting,rgba(0,112,121,1));
  outline-offset: 0px;
}

.c0 > input {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c1 {
  width: 100%;
  border: none;
  background: transparent;
  padding-left: 8px;
  padding-top: 6px;
  padding-right: 8px;
  padding-bottom: 6px;
  font-family: Equinor;
  font-size: 1.000rem;
  font-weight: 400;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  line-height: 1.500em;
  outline: none;
  padding-left: 8px;
  padding-right: 8px;
}

.c1::-webkit-input-placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c1::-moz-placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c1:-ms-input-placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c1::placeholder {
  color: var(--eds_text__static_icons__tertiary,rgba(111,111,111,1));
}

.c1:disabled {
  color: var(--eds-input-color);
  cursor: not-allowed;
}

<div
    class="c0"
    readonly=""
  >
    <input
      class="c1"
      readonly=""
      style="resize: none;"
      type="text"
      value="input value"
    />
  </div>
</DocumentFragment>
`;
